src/optimizer/statistics/stats_storage.cpp

```diff
@@ -6,39 +6,99 @@
-    return common::ManagedPointer<TableStats>(table_it->second);
+  if (table_it == table_stats_storage_.end()) {
+    stats_latch_.UnlockShared();
+    InsertTableStats(database_id, table_id, accessor);
+    table_it = table_stats_storage_.find(stats_storage_key);
+    stats_latch_.LockShared();
   }
-  return common::ManagedPointer<TableStats>(nullptr);
+
+  auto &stats_storage_value = table_it->second;
+
+  UpdateStaleColumns(table_id, &stats_storage_value, accessor);
+
+  stats_latch_.UnlockShared();
+
+  return stats_storage_value;
 }
 
-bool StatsStorage::InsertTableStats(catalog::db_oid_t database_id, catalog::table_oid_t table_id,
-                                    TableStats table_stats) {
-  StatsStorageKey stats_storage_key = std::make_pair(database_id, table_id);
-  auto table_it = table_stats_storage_.find(stats_storage_key);
+/*
+ * Currently when getting statistics for a table we get and cache the statistics for the entire table. If we find that
+ * large tables are filling up the cache when we only need some of the columns, we may want to revisit this so we only
+ * cache the columns we use.
+ */
+std::unique_ptr<TableStats> StatsStorage::GetTableStats(const catalog::db_oid_t database_id,
+                                                        const catalog::table_oid_t table_id,
+                                                        catalog::CatalogAccessor *accessor) {
+  auto &stats_storage_value = GetStatsStorageValue(database_id, table_id, accessor);
+  common::SharedLatch::ScopedSharedLatch table_latch{&stats_storage_value.shared_latch_};
+  return stats_storage_value.table_stats_.Copy();
+}
+
+/*
+ * Currently when getting the statistics for a column we get and cache the statistics for the entire table. If we find
+ * that large tables are filling up the cache, we may want to revisit this so we only cache the column requested.
+ */
+std::unique_ptr<ColumnStatsBase> StatsStorage::GetColumnStats(catalog::db_oid_t database_id,
+                                                              catalog::table_oid_t table_id,
+                                                              catalog::col_oid_t column_oid,
+                                                              catalog::CatalogAccessor *accessor) {
+  auto &stats_storage_value = GetStatsStorageValue(database_id, table_id, accessor);
+  common::SharedLatch::ScopedSharedLatch table_latch{&stats_storage_value.shared_latch_};
+  NOISEPAGE_ASSERT(stats_storage_value.table_stats_.HasColumnStats(column_oid), "Should have stats for all columns");
+  return stats_storage_value.table_stats_.GetColumnStats(column_oid)->Copy();
+}
 
-  if (table_it != table_stats_storage_.end()) {
-    OPTIMIZER_LOG_TRACE("There already exists a TableStats object with the given oids.");
-    return false;
+void StatsStorage::MarkStatsStale(catalog::db_oid_t database_id, catalog::table_oid_t table_id,
+                                  const std::vector<catalog::col_oid_t> &col_ids) {
```

Comment:
I think you need `stats_latch_` in this function as well, since it reads from `table_stats_storage_`.

Reply:
Done